protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+package protofsm
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+const (
+	// pollInterval is the interval at which we'll poll the SendWhen
+	// predicate if specified.
+	pollInterval = time.Millisecond * 100
+)
+
+// EmitedEvent is a special type that can be emitted by a state transition.
+// This can container internal events which are to be routed back to the state,
+// or external events which are to be sent to the daemon.
+type EmittedEvent[Event any] struct {
+	// InternalEvent is an optional internal event that is to be routed
+	// back to the target state. This enables state to trigger one or many
+	// state transitions without a new external event.
+	InternalEvent fn.Option[Event]
+
+	// ExternalEvent is an optional external event that is to be sent to
+	// the daemon for dispatch. Usually, this is some form of I/O.
+	ExternalEvents fn.Option[DaemonEventSet]
+}
+
+// StateTransition is a state transition type. It denotes the next state to go
+// to, and also the set of events to emit.
+type StateTransition[Event any, Env Environment] struct {
+	// NextState is the next state to transition to.
+	NextState State[Event, Env]
+
+	// NewEvents is the set of events to emit.
+	NewEvents fn.Option[EmittedEvent[Event]]
+}
+
+// Environment is an abstract interface that represents the environment that
+// the state machine will execute using. From the PoV of the main state machine
+// executor, we just care about being able to clean up any resources that were
+// allocated by the environment.
+type Environment interface {
+	// CleanUp is a method that'll be called once the state machine has
+	// reached a terminal state. It signals the end of the execution of the
+	// state machine.
+	CleanUp() error
+
+	// TODO(roasbeef): also add checkpointing?
+}
+
+// State defines an abstract state along, namely its state transition function
+// that takes as input an event and an environment, and returns a state
+// transition (next state, and set of events to emit). As state can also either
+// be terminal, or not, a terminal event causes state execution to halt.
+type State[Event any, Env Environment] interface {
+	// ProcessEvent takes an event and an environment, and returns a new
+	// state transition. This will be iteratively called until either a
+	// terminal state is reached, or no further internal events are
+	// emitted.
+	ProcessEvent(event Event, env Env) (*StateTransition[Event, Env], error)
+
+	// IsTerminal returns true if this state is terminal, and false otherwise.
+	IsTerminal() bool
```

Comment:
Totally. Here's an instance of that in the RBF coop PR: https://github.com/lightningnetwork/lnd/blob/43386d5643f961f948bf95513933c7d5a72fc74e/lnwallet/chancloser/rbf_coop_transitions.go#L917-L936

Available tooling wise, I think the best way for us to enforce this would be at the unit test level. I can think of some more involved mechanisms that involve stuff like registering all types in a global map, to then run a generic unit test again, but not so sure we should reach for that yet at this stage.